open_seq2seq/parts/transformer/attention_layer.py

```diff
@@ -152,32 +159,56 @@ def call(self, x, y, bias, cache=None):
         # downcast softmax output
         weights = tf.cast(weights, dtype=dtype)
       else:
+        # Logits shape: [batch, head, decoder, encoder]
+        # Bias shape:   [batch, 1, 1, encoder]
+
+        # Force monotonic attention during inference
+        if positions is not None and self.window_size is not None:
+          assert self.back_step_size is not None
+
+          max_length = tf.shape(logits)[-1]
+
+          # Allow to make back_step_size steps back
+          window_pos = tf.maximum(positions - self.back_step_size, tf.zeros_like(positions))
+
+          # Create attention mask
+          mask_large = tf.sequence_mask(window_pos + self.window_size, maxlen=max_length)
+          mask_large = tf.cast(mask_large, tf.float32)
+          mask_small = tf.sequence_mask(window_pos, maxlen=max_length)
+          mask_small = tf.cast(mask_small, tf.float32)
+          mask = mask_large - mask_small
+          mask = -1e9 * (1 - mask)
+
+          bias = mask + bias
+
+        # Clipping
+        bias = tf.maximum(bias, -1e9)
```

Comment:
What is this clipping for?

Reply:
I sum up a few biases. Don't want them to be a large negatives numbers.